Source/FluentSpreadsheets/Sheets/VisitorImplementations/SheetSegmentVisitor.cs

```diff
@@ -0,0 +1,174 @@
+using FluentSpreadsheets.Sheets.Segments;
+using FluentSpreadsheets.Sheets.Visitors;
+using FluentSpreadsheets.Tools;
+
+namespace FluentSpreadsheets.Sheets;
+
+internal class SheetSegmentVisitor<THeaderData, TRowData, TFooterData> :
+    IPrototypeHeaderSegmentVisitor<THeaderData>
+{
+    private readonly THeaderData _headerData;
+    private readonly IReadOnlyCollection<TRowData> _rowData;
+    private readonly TFooterData _footerData;
+    private SheetSegmentModel? _segment;
+
+    public SheetSegmentVisitor(
+        THeaderData headerData,
+        IReadOnlyCollection<TRowData> rowData,
+        TFooterData footerData)
+    {
+        _headerData = headerData;
+        _rowData = rowData;
+        _footerData = footerData;
+    }
+
+    public SheetSegmentModel BuildSheetSegment(ISheetSegment builder)
+    {
+        builder.Accept(this);
+
+        _segment ??= Visit(builder);
+
+        if (builder is IPrototypeSegmentHeaderCustomizer<THeaderData> headerCustomizer && _segment.HeaderComponent is not null)
+        {
+            var width = _segment.HeaderComponent.Size.Width;
+            var customizedHeader = headerCustomizer.CustomizeHeader(_segment.HeaderComponent, _headerData);
+
+            if (!customizedHeader.Size.Width.Equals(width))
+            {
+                const string message = "The header customizer must not change the width of the header component.";
+                throw new InvalidStructureException(message);
+            }
+            
+            _segment.HeaderComponent = customizedHeader;
+        }
+
+        var segment = _segment;
+        _segment = null;
+
+        return segment;
+    }
+
+    public void Visit<TDestination>(IPrototypeHeaderSegment<THeaderData, TDestination> builder)
+    {
+        TDestination[] data = builder.Select(_headerData).ToArray();
+
+        var headerVisitor = new PrototypeHeaderSegmentVisitor<TDestination>(data);
+        var rowVisitor = new PrototypeRowSegmentVisitor<TDestination, TRowData>(data, _rowData);
+        var footerVisitor = new PrototypeFooterSegmentVisitor<TDestination, TFooterData>(data, _footerData);
+
+        builder.Accept(headerVisitor);
+        builder.Accept(rowVisitor);
+        builder.Accept(footerVisitor);
+
+        IList<IComponent> headerComponents = headerVisitor.Components;
+        IList<IList<IComponent>> rowComponents = rowVisitor.Components;
+        IList<IComponent> footerComponents = footerVisitor.Components;
+
+        ScalePrototypedComponents(data, headerComponents, rowComponents, footerComponents);
+
+        IComponent? header = null;
+        IList<IComponent> rows = Array.Empty<IComponent>();
+        IComponent? footer = null;
+
+        if (headerComponents.Count is not 0)
+            header = new HStackComponent(headerComponents);
+
+        if (rowComponents.Count is not 0)
+        {
+            var convertedRows = ConvertRowsFromSegmentGroupingToRowGrouping(rowComponents, _rowData.Count);
```

Comment:
I think function name is overcompicated here

Reply:
I guess it can be something like TransposeRows

Reply:
For me Transpose is more of a matrix stuff

Reply:
`IList<IList<T>>` is kinda a matrix and that function really transposes the rows :)

Reply:
Oh, my bad, I havent figured out what the function is doing exactly, so yeah, this name will be better